### PR TITLE
fix(birthday): improved date parsing and scheduling

### LINKED ISSUE #1638: bug: improve Date parsing
**Describe the bug**

Date parsing is often, erroneous, `1990/10/12` is parsed and processed as `1990/10/11`.

**To Reproduce**

1. `s!set-birthday 1990/10/12`
2. See that the birthday was set on 11th October instead of 12th.

**Expected behavior**

The date should be parsed correctly, without days off.

**Screenshots**

N/a.

**Additional context**

N/a.

### LINKED ISSUE #1641: request: optional year for birthday
**Is your feature request related to a problem? Please describe.**

While most users will willingly share their birthday, many won't share the year they were born.

**Describe the solution you'd like**

The year in the date parse must be optional, and if not given, the year should be `null` and displayed as `Unknown` in both `{age}` and `{age.ordinal}`.

**Describe alternatives you've considered**

N/a.

**Additional context**

N/a.

### PR DESCRIPTION
**Pull request by kyranet**
Fixes #1641
Fixes #1638

**TODO**:

- [x] Fix compiler errors
- [x] Add timezone compensation (half way #1640)
- [x] Add unit tests
- [x] Add docs
- [x] Add DB migration
- [x] Remove left-over code

### CODE REVIEW

File: src/commands/Birthday/set-birthday.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	runIn: ['text']
+})
+export class UserCommand extends SkyraCommand {
+	public async run(message: GuildMessage, args: SkyraCommand.Args, context: CommandContext) {
+		const [birthdayRole, birthdayChannel, birthdayMessage] = await message.guild.readSettings([
+			Birthday.Role,
+			Birthday.Channel,
+			Birthday.Message
+		]);
+		if (this.isCorrectlyConfigured(birthdayRole, birthdayChannel, birthdayMessage))
```
Comment (vladfrangu) [inline src/commands/Birthday/set-birthday.ts]: So if it's correctly configured.. throw?

File: src/commands/Birthday/set-birthday.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+		return { year: birthDate.getUTCFullYear(), month: birthDate.getUTCMonth() + 1, day: birthDate.getUTCDate() };
+	}
+
+	private isCorrectlyConfigured(birthdayRole: string | Nullish, birthdayChannel: string | Nullish, birthdayMessage: string | Nullish) {
+		// A birthday role, or a channel and message must be configured:
+		return !isNullish(birthdayRole) || (!isNullish(birthdayChannel) && !isNullish(birthdayMessage));
+	}
+
+	private static readonly currentYear = new Date().getUTCFullYear();
```
Comment (vladfrangu) [inline src/commands/Birthday/set-birthday.ts]: This as a class field is.. not ok. What if someone runs the command on Dec 31st 2021 at 23:59 and then again at Jan 1st 00:00

File: src/lib/birthday/types.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+export interface DateWithOptionalYear {
+	year: number | null;
+	month: Month;
+	day: number;
+}
+
+export type BirthDayScheduleEntity = ScheduleEntity & { data: TaskBirthDayData };
+
+export interface TaskBirthDayData extends DateWithOptionalYear, Record<string, unknown> {
```
Comment (vladfrangu) [inline src/lib/birthday/types.ts]:
```suggestion
export type BirthdayScheduleEntity = ScheduleEntity & { data: TaskBirthdayData };

export interface TaskBirthdayData extends DateWithOptionalYear, Record<string, unknown> {
```

File: src/lib/birthday/utils.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+		default:
+			return day <= 30;
+	}
+}
+
+export function yearIsLeap(year: number) {
+	return (year % 4 === 0 && year % 100 !== 0) || year % 400 === 0;
+}
+
+export function isBirthDayTask(task: ScheduleEntity): task is BirthDayScheduleEntity {
```
Comment (vladfrangu) [inline src/lib/birthday/utils.ts]: boi...what is this capitalization??

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: src/commands/Birthday/set-birthday.ts (hunk 2 above)
Comment (kyranet) [reply src/commands/Birthday/set-birthday.ts]: We reload the command. Either way, we have 13 years of margin before we have to accept newer years.

**Review by favna — APPROVED** [review]

**Review by vladfrangu — APPROVED** [review]
LGTM, just... are you removing the birthday role once the day passes? Cause thats how most bots do it :3

**Pull request merged**